templates/activity-stream/impression-stats/impression-stats.1.schema.json

```diff
@@ -0,0 +1,87 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "title": "impression-stats",
+  "properties": {
+    "action": {
+      "type": "string",
+      "pattern": "^activity_stream_impression_stats$"
+    },
+    "client_id": {
+      "type": "string",
+      "pattern": "^n/a$"
+    },
+    "session_id": {
+      "type": "string",
+      "pattern": "^n/a$"
+    },
+    "impression_id": {
+      "type": "string",
+      "pattern": "^\\{[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\\}$"
+    },
+    "addon_version": {
+      "type": "string"
+    },
+    "version": {
+      "type": "string"
+    },
+    "release_channel": {
+      "type": "string"
+    },
+    "locale": {
+      "type": "string"
+    },
+    "topic": {
+      "type": "string"
+    },
+    "source": {
+      "type": "string"
+    },
+    "page": {
+      "type": "string",
+      "enum": [ "about:newtab", "about:home", "about:welcome", "unknown" ]
+    },
+    "region": {
+      "type": "string"
+    },
+    "profile_creation_date": {
+      "type": "integer"
+    },
+    "user_prefs": {
+      "type": "integer"
+    },
+    "click": {
+      "type": "integer"
+    },
+    "block": {
+      "type": "integer"
+    },
+    "pocket": {
+      "type": "integer"
+    },
+    "tiles": {
```

Comment:
There are some extra complexities about this `tiles` field. See more details [here](https://github.com/mozilla/infernyx#payloads-from-firefox)

In a nutshell, both impression ping and other user interaction pings (click, delete, add to pocket) reuse this structure. For example, an impression ping with 3 Pocket stories,

```js
{
  // ... other fields
  "tiles": [
    {"id": 1000, "pos": 0},
    {"id": 1001, "pos": 1},
    {"id": 1002, "pos": 2},
  ]
}
```  

A click on the Pocket story "1002" at the position "2" would generate the following ping:
```js
{
  // ... other fields
  "click": 0,    // 0 indicating the index of the clicked Pocket story stored in the "tiles" array below 
  "tiles": [
    {"id": 1002, "pos": 2},
  ]
}
```
Infernyx (ETL jobs) will store this ping to the database

```js
{
  "tile_id": 1001,
  "position": 2,
  "clicks": 1,
  "blocked": 0,
  "pinned": 0,
  "pocketed": 0 
}
```

The same rule could be applied to `delete`, `pin`, `save to pocket` etc.

This flat schema is very friendly for the aggregate queries for Pocket, such as counting impressions for each story, clicks, deletes, save to pockets etc. Also, it's quite straightforward to query the unique audience of the certain `tile_id`. Wonder if we can still keep those merits if we're looking to keep this `tiles` as an array in the table.

If not, we might want to split this array either on the client side, or somewhere in the data pipeline.

Reply:
> If not, we might want to split this array either on the client side, or somewhere in the data pipeline.

The model we have in mind for the data pipeline in GCP is that every schema in mozilla-pipeline-schemas will map to exactly one table in BigQuery. The table in BigQuery will have exactly the same structure as the payload, save for some additional metadata fields.

So, we'd still want to define some sort of ETL process to serve the kinds of needs you're talking about here.

This kind of transformation sounds like a good candidate for either an on-demand SQL view or a "materialized view" (called a "destination table" in BigQuery) where we have a query run on some schedule to update the target.

If latency is a bigger issue, we could also consider a dataflow streaming job to do such a transformation.

Reply:
> The model we have in mind for the data pipeline in GCP is that every schema in mozilla-pipeline-schemas will map to exactly one table in BigQuery. The table in BigQuery will have exactly the same structure as the payload, save for some additional metadata fields.

Gotcha. Since BigQuery has a great query support on Array types, I think we can just leave the schema as-is, and try it out without introducing other complexities for now.